### PR TITLE
feat: Add option to select mysql dbs to back up

### PR DESCRIPTION
**Pull request by angonz**
Add BACKUP_MYSQL_DATABASES setting to list MySQL databases to back up.

### CODE REVIEW

File: tutorbackup/templates/backup/build/backup/backup_services.py [hunk 1]
```diff
@@ -45,9 +45,16 @@ def mysqldump():
     port = ENV['MYSQL_PORT']
     outfile = MYSQL_DUMPFILE
 
+    mysql_databases = ENV.get('MYSQL_DATABASES')
+    if mysql_databases:
+        databases_statement = f"--databases {mysql_databases}"
```
Comment (foadlind) [inline tutorbackup/templates/backup/build/backup/backup_services.py:50]: `mysqldump --databases 'db_1 db_2'` breaks. You need to remove the single quotes from the `mysql_databases` string. It should result in `mysqldump --databases db_1 db_2`

**Review by foadlind — CHANGES_REQUESTED** [review]
Thanks @angonz for this PR! In general it looks good. However, I ran into some issues while testing this locally:

- This breaks if the user has not set `BACKUP_MYSQL_DATABASES` in their `config.yaml`. You need to add `MYSQL_DATABASES` to `config["defaults"]` in [plugin.py](https://github.com/hastexo/tutor-contrib-backup/blob/main/tutorbackup/plugin.py#L14). Please set the default value to an empty string.
- See inline for the issue with multiple database names

**Comment by fghaas** [comment]
> * This breaks if the user has not set `BACKUP_MYSQL_DATABASES` in their `config.yaml`. You need to add `MYSQL_DATABASES` to `config["defaults"]` in [plugin.py](https://github.com/hastexo/tutor-contrib-backup/blob/main/tutorbackup/plugin.py#L14). Please set the default value to an empty string.

Shouldn't the default be an empty list?

(It won't functionally matter much now, but it will if we ever add type-checking to this plugin like Tutor already does.)

**Comment by foadlind** [comment]
Yes that's a good point. Empty list is better.

**Comment by angonz** [comment]
Is it really necessary? I am checking if the setting exixts before assigning the env var, and ENV.get() should return None if it is not set.

Saludos,
Andrés

El 26 ago. 2022, a la(s) 05:33, Foad Lind ***@***.***> escribió:

﻿

Yes that's a good point. Empty list is better.

**Comment by fghaas** [comment]
> Is it really necessary? I am checking if the setting exixts before assigning the env var, and ENV.get() should return None if it is not set.

It is necessary. `ENV.get()` may work, but the Jinja test `{% if BACKUP_MYSQL_DATABASES %}` fails if the variable is unset.

**Comment by angonz** [comment]
I see... I should have checked by `is defined` in the template. Anyway, I added the default as empty list and remove the quotes of the docker-compose file. Hope it's better now...

File: tutorbackup/patches/k8s-jobs [hunk 2]
```diff
@@ -31,6 +31,10 @@ spec:
               value: '{{ MYSQL_ROOT_USERNAME }}'
             - name: MYSQL_ROOT_PASSWORD
               value: '{{ MYSQL_ROOT_PASSWORD }}'
+            {% if BACKUP_MYSQL_DATABASES %}
+            - name: MYSQL_DATABASES
+              value: '{{ BACKUP_MYSQL_DATABASES | join(" ") }}'
```
Comment (foadlind) [inline tutorbackup/patches/k8s-jobs:36]: The single quotes here should also be removed.

File: tutorbackup/patches/k8s-jobs [hunk 3]
```diff
@@ -104,6 +108,10 @@ spec:
                   value: '{{ MYSQL_ROOT_USERNAME }}'
                 - name: MYSQL_ROOT_PASSWORD
                   value: '{{ MYSQL_ROOT_PASSWORD }}'
+                {% if BACKUP_MYSQL_DATABASES %}
+                - name: MYSQL_DATABASES
+                  value: '{{ BACKUP_MYSQL_DATABASES | join(" ") }}'
```
Comment (foadlind) [inline tutorbackup/patches/k8s-jobs:113]: same as above

File: tutorbackup/patches/k8s-jobs [hunk 4]
```diff
@@ -179,6 +187,10 @@ spec:
...
```
Comment (foadlind) [inline tutorbackup/patches/k8s-jobs:192]: same as above

**Review by foadlind — CHANGES_REQUESTED** [review]
I think you should remove the single quotes in k8s-jobs as well otherwise it will create the same problem.

File: tutorbackup/patches/k8s-jobs (hunk 4 above)
Comment (angonz) [reply tutorbackup/patches/k8s-jobs:192]: I saw those quotes weren't translated to the settings, but anyway I removed them

File: tutorbackup/patches/k8s-jobs (hunk 2 above)
Comment (angonz) [reply tutorbackup/patches/k8s-jobs:36]: done

File: tutorbackup/templates/backup/build/backup/backup_services.py (hunk 1 above)
Comment (angonz) [reply tutorbackup/templates/backup/build/backup/backup_services.py:50]: done

**Review by fghaas — COMMENTED** [review]
Unless @foadlind has any remaining objections, this looks good to me now. @angonz, could you please squash the three commits into one and then force-push to your branch, so we can merge this as one commit?

File: tutorbackup/patches/k8s-jobs (hunk 2 above)
Comment (foadlind) [reply tutorbackup/patches/k8s-jobs:36]: It's still there though.

File: tutorbackup/patches/k8s-jobs (hunk 4 above)
Comment (foadlind) [reply tutorbackup/patches/k8s-jobs:192]: It's still there.

**Review by foadlind — CHANGES_REQUESTED** [review]
@angonz please remove the two remaining instances of `'{{ BACKUP_MYSQL_DATABASES | join(" ") }}'` in k8s-jobs.

**Comment by angonz** [comment]
Sorry for that... I think now it's ok

**Review by foadlind — APPROVED** [review]
This looks good to me now. Thanks @angonz!

**Pull request merged**